src/coreclr/jit/codegenxarch.cpp

```diff
@@ -6074,17 +6074,31 @@ void CodeGen::genCall(GenTreeCall* call)
     }
 #endif // defined(DEBUG) && defined(TARGET_X86)
 
-    // When it's a PInvoke call and the call type is USER function, we issue VZEROUPPER here
-    // if the function contains 256bit AVX instructions, this is to avoid AVX-256 to Legacy SSE
-    // transition penalty, assuming the user function contains legacy SSE instruction.
-    // To limit code size increase impact: we only issue VZEROUPPER before PInvoke call, not issue
-    // VZEROUPPER after PInvoke call because transition penalty from legacy SSE to AVX only happens
-    // when there's preceding 256-bit AVX to legacy SSE transition penalty.
-    // This applies to 512bit AVX512 instructions as well.
-    if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) && (GetEmitter()->Contains256bitOrMoreAVX()))
-    {
-        assert(compiler->canUseVexEncoding());
-        instGen(INS_vzeroupper);
+    if (compiler->canUseVexEncoding())
+    {
+        // The Intel optimization manual guidance in `3.11.5.3 Fixing Instruction Slowdowns` states:
+        //   Insert a VZEROUPPER to tell the hardware that the state of the higher registers is clean
+        //   between the VEX and the legacy SSE instructions. Often the best way to do this is to insert a
+        //   VZEROUPPER before returning from any function that uses VEX (that does not produce a VEX
+        //   register) and before any call to an unknown function.
+
+        bool needsZeroupper = false;
+
+        // TODO-XArch-CQ: We should emit vzeroupper before R2R calls if they aren't known to require AVX+
+        // TODO-XArch-CQ: We should emit vzeroupper before JIT helpers known to use floating-point
+
+        if (call->IsPInvoke() && (call->gtCallType != CT_HELPER))
+        {
+            // Since P/Invokes are not compiled by the runtime, they are typically "unknown" since they
+            // may use the legacy encoding.
+
+            needsZeroupper = true;
+        }
+
+        if (needsZeroupper)
+        {
+            instGen(INS_vzeroupper);
+        }
```

Comment:
CC. @jkotas, @jkoritzinsky

Would appreciate some input on this. We definitely want to emit this before any P/Invoke to a user-defined function (`CT_USER_FUNC`), even if the method wasn't `Contains256BitOrMoreAvx` itself (as it may have been called by a function which itself used such instructions).

I believe this also means we want to do this for `CT_INDIRECT`, since we know the indirection is to native.

We would also, presumably, want to do this for `CT_HELPER` for cases like the floating-point conversion helpers or the Math ecalls. However, I'm not quite sure the "best" way to check for that here.

Likewise, I expect we want to explicitly do this for any calls to `R2R` functions where we know they weren't compiled for AVX+. However, I'm also unsure the "best" way to check for that scenario.

Any input would be appreciated.

Reply:
> for cases like the floating-point conversion helpers or the Math ecalls. However, I'm not quite sure the "best" way to check for that here.

I think you can check for `CORINFO_FLG_NOGCCHECK` with floating point argument. `CORINFO_FLG_NOGCCHECK` indicates that the method is FCall. Using a flag named NOGCCHECK to check for an FCall is a bit of a hack. If you go with it, we may want to rename the flag to `CORINFO_FLG_INTERNAL_CALL` or something like that.

Reply:
I just pushed up a bit that does it a little differently.

We have to morph the args for every call to determine ABI information and this requires us already checking if something might exist in float/simd register or not. So I used one of the several free bits in `CallArgs` to track if any parameter was a SIMD or floating-point type, which should make it very low cost and is relevant information that can be potentially used for other future optimizations.

This should give us correct handling for everything except for:
* any implicit SIMD usage that a native compiler might opt to do in a helper
* Calls to R2R methods which use floating-point
  * This should only impact T0 code
  * It might be worth having the dispatch stub handle the vzeroupper in this case, so it can be removed for T1; but we can always look at that separately since its not a net new scenario.